_posts/2022-10-27-using-kani-with-the-bolero-property-testing-framework.md

```diff
@@ -0,0 +1,311 @@
+```
+
+Notice this line:
+
+```rust
+#[cfg_attr(kani, kani::proof)]
+```
+
+whose purpose is to treat this function as a proof harness when run with Kani, but treat it as an ordinary test otherwise.
+
+The Bolero harness generates arrays of length `N` (set to 5 in this example), and an index, calls `nth_rev`, and checks that it returns `Some` if the index it received was less than the array length.
+
+Let's first test this with fuzzing using Bolero's default engine (`libfuzzer`):[^footnote-dev-dep]:
+
+```bash
+$ cargo bolero test check_rev
+# snip ...
+test failed; shrinking input...
+
+======================== Test Failure ========================
+
+Input: 
+(
+    [
+        0,
+        0,
+        0,
+        0,
+        0,
+    ],
+    0,
+)
+
+Error: 
+panicked at 'index out of bounds: the len is 5 but the index is 5', src/main.rs:6:21
+```
+
+The fuzzing engine found an input that causes the expected out-of-bounds error: an array of 5 zeros and an index of zero. 
+
+Let's see if Kani finds the same issue:
+
+```bash
+$ cargo bolero test check_rev --engine kani
+# snip ...
+SUMMARY:
+ ** 1 of 337 failed (1 unreachable)
+Failed Checks: index out of bounds: the length is less than or equal to the given index
+ File: "/home/ubuntu/examples/demos/rev-bolero/src/main.rs", line 6, in nth_rev
+```
+
+It did! Now, let's fix the bug by updating the buggy line to:
+
+```rust
+        let rev_index = arr.len() - index - 1;
+```
+
+and re-run the fuzzer:
+
+```
+$ cargo bolero test check_rev
+# snip ...
+#3773   REDUCE cov: 135 ft: 138 corp: 18/282b lim: 52 exec/s: 0 rss: 43Mb L: 8/32 MS: 3 ChangeByte-ChangeBinInt-EraseBytes-
+#262144 pulse  cov: 135 ft: 138 corp: 18/282b lim: 2620 exec/s: 131072 rss: 68Mb
+#524288 pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 131072 rss: 93Mb
+#1048576        pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 116508 rss: 142Mb
+```
+
+The fuzzer will run for a while without finding any failures. We can now run Kani to *prove* the assertion in our harness:
+
+```
+$ cargo bolero test check_rev --engine kani
+# snip ...
+SUMMARY:
+ ** 0 of 338 failed (1 unreachable)
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.90419155s
+```
+
+and voila! Kani managed to prove that there none of the checks can fail for any values of the inputs.
+
+## If we have proof capability, do we still need fuzzing?
+
+One question that might arise is: is there any point in using fuzzing alongside with Kani?
+In other words, if we can prove a harness for all inputs with Kani, would we still want to use fuzzing?
+The answer is yes, because fuzzing and Kani provide complementary benefits.
+Kani verifies a Rust program by symbolically analyzing its code.
+This allows Kani to make mathematical statements about the expected semantics of the Rust code being verified.
+Kani allows you to prove that, **for all possible inputs**, the code under verification follows its specification, **assuming everything else functions correctly**.  
```

Comment:
What does "assuming everything else functions correctly" mean here? I think this might be a bit too loose.

Reply:
I expanded on that a bit.